this closes #103 issue

jobol opened the pull request:
This closes #103 issue.

This implements recursion for chsmack and also improves how to delete smack properties.

Briefly, it is now possible to set properties while resetting other properties in the same time.

Example:

    chsmack -a User -T dir

set the label of the directory 'dir' and remove its transmute property (if any)

jobol commented:
ping dring

rafal-krypa commented:
Hi. Thank you for the patches and sorry that it took so long.
I am reviewing your code right now and will post some comments soon.

jobol commented:
@rafal-krypa had you time to review?

jobol commented:
ping

rafal-krypa commented:
I have finally managed to through your code. Great work!

After so long time I feel obliged to merge your changes without nit-picking and eventually fix minor issues later. Nevertheless if you would be willing to shortly discuss few aspects before we proceed, I'd be obliged.

1. I would prefer to have your commits without Gerrit Change-Id line. It is currently a policy in many open-source projects to treat Change-Ids as internal stuff that shouldn't be put in public code repositories.
2. Libsmack is written using Linux kernel-ish coding style. Your changes don't obey those rules in few places, mostly about having "else" on the same line as the preceding closing brace. If you are willing, please consult your patches with checkpatch.pl script from Linux kernel. Alternatively I will address those issues in a later commit, after merging yours.
3. I found some minor issues in the code that I'd like to comment in-line.

Let's stay in touch and merge your work.

rafal-krypa commented on utils/chsmack.c:
```diff
@@ -81,6 +81,55 @@ static struct option *option_by_char(int car)
+	}
+}
+
+/* modify transmutation of a file */
```
of a directory

rafal-krypa commented on doc/chsmack.8:
```diff
@@ -139,6 +139,13 @@ while still ensuring that the config files of the editor remain protected.
+.TP
+.B -D, --drop
+
+Use this option to remove the smack properties that are not set for files.
```
This description is confusing. "Properties that are not set" can be understood in the chsmack invocation context, as you probably intended, or in the existence of Smack attributes in file system. The latter interpretation raises question - why would I drop a property that is already dropped?

How about an alternative description:
> Use this option to instruct chsmack to remove all attributes that aren't set by other flags. For example chsmack -a Foo -D would drop all attributes except security.SMACK64.

rafal-krypa replied:
Now I see that you have added an example below. In that case, please skip the part of my comment about example.

rafal-krypa commented on doc/chsmack.8:
```diff
@@ -26,7 +26,7 @@ chsmack \- Change or list the Smack properties of filesystem objects
 
 .B chsmack [-L|--dereference] [--] files...
 
-.B chsmack [-L|--dereference] [-D|--drop] props... [--] files...
```
Why remove it?

rafal-krypa commented on doc/chsmack.8:
```diff
@@ -26,7 +26,7 @@ chsmack \- Change or list the Smack properties of filesystem objects
 .B chsmack [-L|--dereference] [--] files...
 
-.B chsmack [-L|--dereference] props... [--] files...
+.B chsmack [-L|--dereference] [-D|--drop] props... [--] files...
```
With the additional drop options that you introduce in the next patch, is there really a need for drop-all option added here? The example from the commit message could be rewritten as:
> chsmack -a User::Item -EMT

I'm in favor of the selective removal options, with this -D thing the user doesn't have to be explicit about the attributes that are supposed to be removed.

jobol replied:
For sure. But why to not keep that facility? It is good to write "chsmack -D file" it drops all.

rafal-krypa commented on utils/chsmack.c:
```diff
@@ -148,6 +148,59 @@ static void modify_transmute(const char *path)
+	printf("%s", path);
+
+	errno = 0;
+	rc = (int)smack_new_label_from_path(path,
```
Is downcasting from `ssize_t` to `int` really necessary? Since all `smack_new_label_*` functions return `ssize_t`, changing type of `rc` to `ssize_t` would be better.

rafal-krypa commented on utils/chsmack.c:
```diff
@@ -36,20 +36,21 @@
 	" -E --drop-exec       remove "XATTR_NAME_SMACKEXEC"\n"
 	" -M --drop-mmap       remove "XATTR_NAME_SMACKMMAP"\n"
 	" -T --drop-transmute  remove "XATTR_NAME_SMACKTRANSMUTE"\n"
+	"Obsolete option:\n"
```
The `-d` was made obsolete in another, latter commit.

rafal-krypa commented on utils/chsmack.c:
```diff
@@ -125,12 +130,15 @@ static void modify_transmute(const char *path)
 		else if (!S_ISDIR(st.st_mode)) {
-			fprintf(stderr, "%s: transmute: not a directory\n",
-				path);
+			if (!recursive_flag) {
```
Why not report an error on invocation like:
> chsmack --transmute --recureive some_file
?

You would warn on setting transmute for a file, but addition of --recursive flag changes the error to non-error?

jobol replied:
reason is to keep it simple when "chsmack --transmute --recursive --access LABEL file1 file2 dir1 dir2"

rafal-krypa commented on utils/chsmack.c:
```diff
@@ -162,6 +170,7 @@ static void print_file(const char *path)
 {
 	int rc;
 	char *label;
+	int n = 0;
```
Descriptive variable names are much better.

rafal-krypa commented on utils/chsmack.c:
```diff
@@ -196,9 +205,75 @@ static void print_file(const char *path)
+	struct stat st;
+	int rc;
+	char *file;
+	size_t last, length, l;
```
One more non-descriptive variable. This name is also bad because it looks like 1.

rafal-krypa commented on utils/chsmack.c:
```diff
@@ -196,9 +205,75 @@ static void print_file(const char *path)
+	}
+	for(;;) {
+		rc = readdir_r(dir, &dent, &pent);
+		if (rc != 0 || pent == NULL) {
```
No handling of errors returned from `readdir_r`?

rafal-krypa commented on utils/chsmack.c:
```diff
@@ -196,9 +205,75 @@ static void print_file(const char *path)
+			return;
+		}
+		l = strlen(dent.d_name);
+		if (l && dent.d_name[0] == '.'
```
How about a less confusing:

    if (!strcmp(dent.d_name, ".") || !strcmp(dent.d_name, ".."))

?

jobol commented:
Rafal, your review was useful as usual. Thanks.

First, I rebased to remove the commit-id.
Then I added 2 more commits to integrate your remarks.

regards

jobol commented:
ping
